### DISCUSSION TITLE
Combining an array of Futures

### DISCUSSION THREAD

**Discussion by LandonSchropp** [fluture-js/Fluture#460] [Q&A] [answered]
Is it possible to combine an array of `Future` objects? I'm looking for a `Future` equivilent of `Promise.all`. I imagine it would look something like this:

``` js
let futures = [ Future.of(1), Future.of(2), Future.of(3) ];
Future.all(futures).fork(console.error, console.log);
//> [1, 2, 3];
```

Thanks!

**Comment by mrosata** [comment, marked as answer]
```javascript
Future.parallel(2, [
  Future.of(1), Future.of(2)
])
    .fork(err, done)
```

_Sent from my Google Pixel using [FastHub](https://play.google.com/store/apps/details?id=com.fastaccess.github)_

**Comment by LandonSchropp** [comment]
Thank you!

**Comment by arsdragonfly** [comment]
What if I want to combine a heterogeneous tuple of futures?

**Comment by Avaq** [comment]
@arsdragonfly I use [`both`](https://github.com/fluture-js/Fluture#both) for doing so. If you need more than two elements, you can nest the tuple.

```js
chain (([a, [b, c]]) => ...)
      (both (eventualA) (both (eventualB) (eventualC)))
```

Alternatively, you can override the type of `parallel` to allow heterogeneous array. I have discussed the type of `parallel` at length in this PR: https://github.com/fluture-js/Fluture/pull/184

I don't want to change it, as it opens a door for the exact type of bugs users are trying to fight with TypeScript. I wish I could provide multiple typings files that users can choose from - a strict version and a loosy version, but the TypeScript maintainers feel this would be doing things in the wrong way.

**Comment by Avaq** [comment]
@arsdragonfly Also note that `lift*` is the perfect abstraction for this. If you have a library that gives you a typed `lift` function, then that's the one to go for:

```js
const f = a => b => c => a + b + c;

lift3 (f) (eventualA) (eventualB) (eventualC)
//or for parallelism:
lift3 (f) (Par (eventualA)) (Par (eventualB)) (Par (eventualC))
```
